Microsoft Entra ID provider (#2390)
oauth2-proxy/oauth2-proxy: docs/docs/configuration/providers/ms_entra_id.md

insecure_oidc_skip_issuer_verification=true
```
`insecure_oidc_skip_issuer_verification` setting is required to disable following checks:
* Startup check for matching issuer URL returned from [discovery document](https://login.microsoftonline.com/common/v2.0/.well-known/openid-configuration) with `oidc_issuer_url` setting. Required, as document's `issuer` field doesn't equal to `https://login.microsoftonline.com/common/v2.0`. See [OIDC Discovery 4.3](https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderConfigurationValidation).
* Matching ID token's `issuer` claim with `oidc_issuer_url` setting during ID token validation. Required to support tokens issued by diffrerent tenants. See [OIDC Core 3.1.3.7](https://openid.net/specs/openid-connect-core-1_0.html#IDTokenValidation). 

To provide additional security against the insecure setting, Entra ID provider performs additional check on `issuer` claim to accept tokens issued only by `https://login.microsoftonline.com/{tenant-id}/v2.0`.

### Example configurations
Single-tenant app without groups (*groups claim* not enabled). Consider using generic OIDC provider:
```shell

> provider="entra-id"

Reviewer: What exactly do you mean?

> provider="entra-id"

Reviewer: We don't really have a naming scheme. Most are just lower case like `github`, `bitbucket`, `google` or `bitbucket` but we have `login.gov` as well.

> provider="entra-id"

Reviewer 2: Perhaps at some point we should standardise these, so they are all, for example, PascalCase or something?

oidc_issuer_url="https://login.microsoftonline.com/<tenant-id>/v2.0"
client_id="<client-id>"
client_secret="<client-secret>"
scope="openid"
```

Single-tenant app with up to 200 groups (*groups claim* enabled). Consider using generic OIDC provider:
```shell
provider="entra-id"
oidc_issuer_url="https://login.microsoftonline.com/<tenant-id>/v2.0"
client_id="<client-id>"
client_secret="<client-secret>"
scope="openid"

> allowed_groups=["ac51800c-2679-4ecb-8130-636380a3b491"]

Reviewer: Yes as per every provider

> allowed_groups=["ac51800c-2679-4ecb-8130-636380a3b491"]

Reviewer 2: In the future, maybe it would be possible/better to have the group names/email addresses? But I guess that doesn't make things unique at the provider side 🤔

---
id: ms_entra_id
title: Microsoft Entra ID
---

> OIDC-compliant provider for Microsoft Entra ID (Azure AD successor) application registrations. This providers support Azure oAuth V2 endpoint only, if you want to use V1 endpoint, try legacy `azure` provider. 

Reviewer 3: ```suggestion
OIDC-compliant provider for Microsoft Entra ID (Azure AD successor) application registrations. This provider supports the Azure OAuth V2 endpoint only, if you want to use the V1 endpoint, try the legacy [azure](azure.md) provider.

## Configure application registration

> To start, create an App registration with minimal permissions, assign redirect URI and generate secret. All account types are supported (Single tenant, multi tenant, multi tenant with MS accounts, MS accounts only).

Reviewer 3: ```suggestion
To start, create an App registration with minimal permissions, assign a redirect URI, and generate a secret. All account types are supported (Single tenant, multi-tenant, multi-tenant with MS accounts, and MS accounts only).

<details>
    <summary>See Azure Portal example</summary>
    <div class="videoBlock">
        <video controls>
            <source src="/oauth2-proxy/azure-videos/create-app-reg.mkv"/>
        </video>
    </div>

When created with Portal, App registration automatically creates a delegated API permission for `User.Read`. 
</details>

<details>
    <summary>See Terraform example</summary>

    resource "azuread_application" "auth" {
        display_name     = "oauth2-proxy"
        sign_in_audience = "AzureADMyOrg" # Others are also supported

        web {
            redirect_uris = [
                "https://podinfo.lakis.tech/oauth2/callback",
            ]
        }

        required_resource_access {
            resource_app_id = "00000003-0000-0000-c000-000000000000" # Microsoft Graph
            resource_access {
                id   = "e1fe6dd8-ba31-4d61-89e7-88639da4683d" # User.Read delegated
                type = "Scope"
            }
        }
    }

    resource "azuread_application_password" "apppass" {
        application_id = azuread_application.auth.id

>     }

Reviewer 3: Does it make sense to put this in a code block? (\```...```)

Author: Right, by some reason docusaurus misbehaved before, with code block everything looks good
![image](https://github.com/oauth2-proxy/oauth2-proxy/assets/12859079/150e247c-ba58-4ff0-8032-a0aa3279c502)

</details>

This is sufficient for a simple authentication scenario.

### Configure `groups` claim

> If you want to make use of groups (for example, use `--allowed-group` feature of oauth2-proxy or authorize based on groups inside your service), you need to configure `groups` claim to be present in the ID token:

Reviewer 3: ```suggestion
If you want to make use of groups (i.e. `--allowed-group` or authorize based on groups inside your service), you need to configure the `groups` claim to be present in the ID token:

<details>
    <summary>See Azure Portal example</summary>
    <div class="videoBlock">
        <video controls>
            <source src="/oauth2-proxy/azure-videos/create-groups-claim.mkv"/>
        </video>
    </div>
</details>
<details>
    <summary>See Terraform example</summary>

    resource "azuread_application" "auth" {
        display_name     = "oauth2-proxy"
        sign_in_audience = "AzureADMyOrg" # Other alre also supported

        group_membership_claims = [
            "SecurityGroup"
        ]
        optional_claims {
            id_token {
                name = "groups"
            }
        }

        web {
            redirect_uris = [
                "https://podinfo.lakis.tech/oauth2/callback",
            ]
        }

        required_resource_access {
            resource_app_id = "00000003-0000-0000-c000-000000000000" # Microsoft Graph
            resource_access {
                id   = "e1fe6dd8-ba31-4d61-89e7-88639da4683d" # User.Read delegated
                type = "Scope"
            }
        }
    }

    resource "azuread_application_password" "apppass" {
        application_id = azuread_application.auth.id

>     }

Reviewer 3: Same as above

</details>

### Configure group overage support

> Azure has a limit of 200 groups in the JWT. If you can't avoid such a bug number and still want to access the groups, you need to grant `GroupMember.Read.All` delegated permission to the app registration so oauth2-proxy can read all the groups from Graph API. **NOTE**: This permission by default requires an admin consent!

Reviewer 3: ```suggestion
Azure has a limit of 200 groups in the JWT. If you can't avoid such a big number and still want to access the groups, you need to grant `GroupMember.Read.All` delegated permission to the app registration so oauth2-proxy can read all the groups from Graph API. **NOTE**: This permission by default requires admin consent!

        </video>
    </div>
</details>
<details>
    <summary>See Terraform example</summary>

    resource "azuread_application" "auth" {
        display_name     = "oauth2-proxy"
        sign_in_audience = "AzureADMyOrg" # Other alre also supported

        group_membership_claims = [
            "SecurityGroup"
        ]
        optional_claims {
            id_token {
                name = "groups"
            }
        }

        web {
            redirect_uris = [
                "https://podinfo.lakis.tech/oauth2/callback",
            ]
        }

        required_resource_access {
            resource_app_id = "00000003-0000-0000-c000-000000000000" # Microsoft Graph
            resource_access {
                id   = "e1fe6dd8-ba31-4d61-89e7-88639da4683d" # User.Read delegated
                type = "Scope"
            }
            
            resource_access {
                id   = "bc024368-1153-4739-b217-4326f2e966d0" # GroupMember.Read.All delegated - admin consent required
                type = "Scope"
            }

        }
    }

    resource "azuread_application_password" "apppass" {
        application_id = azuread_application.auth.id

>     }

Reviewer 3: Same as above

Admin consent is required after creation by Terraform
</details>

## Configure provider

> Provider is OIDC-compliant, so all the OIDC parameters are honoured. Additional provider-specific configuration parameters are:
> * `ms-entra-id-skip-groups-from-graph` - never read groups from Graph API, even when ID token indicates that there's a group overage. Set if you expect group overage in some cases, but still don't want to assign wide `GroupMember.Read.All`. Defaults to `false`. If you don't need groups, consider skipping `groups` claim in the app registration.
> * `ms-entra-id-multi-tenant-allowed-tenant` - speciy list of allowed tenants to be authenticated through multi-tenant app. When not set, all tenants are allowed. Defaults to `[]` (all tenants).

Reviewer 3: ```suggestion
The provider is OIDC-compliant, so all the OIDC parameters are honored. Additional provider-specific configuration parameters are:
* `ms-entra-id-skip-groups-from-graph` - never read groups from Graph API, even when the ID token indicates that there's a group overage. Set if you expect group overage in some cases, but still don't want to assign wide `GroupMember.Read.All`. Defaults to `false`. If you don't need groups, consider skipping the `groups` claim in the app registration.
* `ms-entra-id-multi-tenant-allowed-tenant` - specify a list of allowed tenants to be authenticated through a multi-tenant app. When not set, all tenants are allowed. Defaults to `[]` (all tenants).
```

### Scope

> For Azure-only apps (multi-tenant and single-tenant), the only required oAuth scope is `openid`:

Reviewer 3: ```suggestion
For Azure-only apps (multi-tenant and single-tenant), the only required OAuth scope is `openid`:
```

> ```

Reviewer 3: ```suggestion
```shell
```

- --scope=openid
```
For personal MS accounts, the scope has to be extended with `email` and `profile`:

> ```

Reviewer 3: ```suggestion
```shell
```

- --scope=openid profile email
```

> It's recommended to configure the scopes explicitly, otherwise you may experience issues with allowing groups (Azure doesn't support `groups` scope which is automatically included when you configure allowed groups).

Reviewer 3: ```suggestion
It's recommended to configure the scopes explicitly, otherwise, you may experience issues with allowing groups (Azure doesn't support the `groups` scope, which is automatically included when you configure allowed groups).
```

### Single-tenant
Simple single-tenant configuration:

> ```

Reviewer 3: ```suggestion
```shell
```

- --provider=ms-entra-id
- --oidc-issuer-url=https://login.microsoftonline.com/{tenantId}/v2.0
- --client-id=<valid-client-id>
- --client-secret=<valid-client-secret>
- --scope=openid
```

### Multi-tenant

> Multi-tenant apps require to disable OIDC issuer verification, as `issuer` field in the [discovery document](https://login.microsoftonline.com/common/v2.0/.well-known/openid-configuration) is a template, not an exact value:

Reviewer 3: ```suggestion
Multi-tenant apps require you to disable OIDC issuer verification, as `issuer` field in the [discovery document](https://login.microsoftonline.com/common/v2.0/.well-known/openid-configuration) is a template, not an exact value:

> ```

Reviewer 3: ```suggestion
```shell
```

- --provider=ms-entra-id
- --oidc-issuer-url=https://login.microsoftonline.com/common/v2.0
- --client-id=<valid-client-id>
- --client-secret=<valid-client-secret>
- --insecure-oidc-skip-issuer-verification
- --scope=openid profile email
```

> Configuration above insecurely allows all tenants, to allow specific tenants:

Reviewer 3: ```suggestion
The configuration above insecurely allows all tenants. To allow specific tenants, use the configuration below as an example:

> ```

Reviewer 3: ```suggestion
```shell
```
